### PR TITLE
Fix multigrid segfault errors under UNIX

### PR DESCRIPTION
**Pull request by CKehl**
In our longer running test case, we found out that the field sampling of multigrids still leads to errors (specifically there: Segmentation Faults and ctypes errors, such as:
```
tests/test_grids.py:584: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev192+g74c9143.d20200402-py3.7.egg/parcels/particleset.py:640: in execute
    self.kernel.execute(self, endtime=time, dt=dt, recovery=recovery, output_file=output_file)
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev192+g74c9143.d20200402-py3.7.egg/parcels/kernel.py:409: in execute
    self.execute_jit(pset, endtime, dt)
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev192+g74c9143.d20200402-py3.7.egg/parcels/kernel.py:253: in execute_jit
    particle_data = byref(pset.ctypes_struct)
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev192+g74c9143.d20200402-py3.7.egg/parcels/particleset.py:306: in ctypes_struct
    cdata = [cdata_for(v) for v in self.ptype.variables]
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev192+g74c9143.d20200402-py3.7.egg/parcels/particleset.py:306: in <listcomp>
    cdata = [cdata_for(v) for v in self.ptype.variables]
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev192+g74c9143.d20200402-py3.7.egg/parcels/particleset.py:302: in cdata_for
    cdata_array = np.ctypeslib.as_ctypes(self.particle_data[v.name])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
obj = array([[1, 2, 0],
       [0, 2, 0],
       [0, 0, 0],
       [0, 0, 0]], dtype=int32)
    def as_ctypes(obj):
        """Create and return a ctypes object from a numpy array.  Actually
        anything that exposes the __array_interface__ is accepted."""
        ai = obj.__array_interface__
        if ai["strides"]:
>           raise TypeError("strided arrays not supported")
E           TypeError: strided arrays not supported
/var/scratch/bin-x86_64/anaconda/envs/OpenParcels_MPI/lib/python3.7/site-packages/numpy/ctypeslib.py:531: TypeError
```

based on that, we created an new, expanded test in `tests/test_fieldset_sampling.py` that resembles the error occurence better, called `test_sampling_multigrids_non_vectorfield()`.

specifically, in the custom test kernel, the sampling of the 'B' field fails.

### CODE REVIEW

**Comment by erikvansebille** [comment]
@angus-g, might you be able to help with this bug? It seems to be relates to #782. Interestingly the new unit test does work for 1 particle, but not for multiple particles. Might it be because the new `ParticleSet.particle_data['xi']` is transposed the wrong way?

**Comment by angus-g** [comment]
No problem, I'll take a look.

File: parcels/kernel.py [hunk 1]
```diff
@@ -219,7 +219,7 @@ def load_lib(self):
     def execute_jit(self, pset, endtime, dt):
         """Invokes JIT engine to perform the core update loop"""
         if len(pset) > 0 and pset.particle_data['xi'].ndim == 2:
-            assert pset.fieldset.gridset.size == pset.particle_data['xi'].shape[0], \
+            assert pset.fieldset.gridset.size == pset.particle_data['xi'].shape[1], \
```
Comment (CKehl) [inline parcels/kernel.py:222]: great - so now the order is again: particles['xi'][<particle_index>][<grid_index>] -> I like that a lot!

File: parcels/particleset.py [hunk 2]
```diff
@@ -299,9 +299,9 @@ class CParticles(Structure):
             _fields_ = [(v.name, POINTER(np.ctypeslib.as_ctypes_type(v.dtype))) for v in self.ptype.variables]
 
         def cdata_for(v):
-            cdata_array = np.ctypeslib.as_ctypes(self.particle_data[v.name])
-            # cast to bare pointer
-            return cast(cdata_array, POINTER(np.ctypeslib.as_ctypes_type(v.dtype)))
+            data_flat = self.particle_data[v.name].view()
+            data_flat.shape = -1
+            return np.ctypeslib.as_ctypes(data_flat)
```
Comment (CKehl) [inline parcels/particleset.py:304]: nifty shorthand for that conversion, sweet - I would not have thought of that.

**Pull request merged**